inst/r_interface/model_input.hpp.in

```diff
@@ -0,0 +1,119 @@
+{{header}}
+
+#pragma once
+
+#include <Rcpp.h>
+#include <unsupported/Eigen/CXX11/Tensor>
+
+#include "types.hpp"
+
+template <typename T>
+T* r_data(SEXP x) {
+  static_assert(sizeof(T) == 0, "Only specializations of r_data can be used");
+}
+
+template <>
+double* r_data(SEXP x) {
+  return REAL(x);
+}
+
+template <>
+int * r_data(SEXP x) {
+  return INTEGER(x);
+}
```

Comment:
I suggest moving all the bits that do not respond to templating out of this file. There's quite a few utilities here. Basically anything not involving `{{...}}}` can probably go elsewhere. This will make the templating easier to reason about